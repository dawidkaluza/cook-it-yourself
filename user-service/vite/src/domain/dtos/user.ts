
export type SignInRequest = {
  email: string;
  password: string;
};

export type SignInResponse = {
  redirectUrl: string;
  external: boolean;
};

export type RedirectResponse = {
  redirectUrl?: string;
};