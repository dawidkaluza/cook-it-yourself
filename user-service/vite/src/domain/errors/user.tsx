export class InvalidCredentialsError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidCredentialsError";
  }
}

export class InvalidFieldsError extends Error {
  private readonly _fields: Record<string, string>;

  constructor(fields: Record<string, string>) {
    super("Invalid fields: " + fields);
    this.name = "InvalidFieldsError";
    this._fields = fields;
  }


  get fields(): Record<string, string> {
    return this._fields;
  }
}