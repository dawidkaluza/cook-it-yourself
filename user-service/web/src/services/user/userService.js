import axios from "axios";

const axiosInstance = axios.create({
  baseURL: process.env.USER_SERVICE_BASEURL,
  headers: {
    "Accept": "application/json",
    "Content-Type": "application/json"
  }
});

const buildSuccessResponse = (response) => {
  return {
    success: true,
    response
  };
};

const buildErrorResponse = ({ message = "", fields = {} }) => {
  return {
    success: false,
    response: { message, fields }
  };
};

const validateSignInFields = (fields, errors) => {
  let result = true;

  for (const key in fields) {
    if (!fields[key]) {
      errors[key] = errors[key] ?? "Field must not be empty.";
      result = false;
    }
  }

  return result;
};

const handleRedirect = (redirectUrlAsString) => {
  if (!redirectUrlAsString) {
    return buildSuccessResponse({
      redirectUrl: "",
      external: false
    });
  }

  const redirectUrl = new URL(redirectUrlAsString);
  const userServiceBaseUrl = new URL(process.env.USER_SERVICE_BASEURL);

  switch (redirectUrl.host) {
    case window.location.host: {
      return buildSuccessResponse({
        redirectUrl: redirectUrl.pathname + redirectUrl.search,
        external: false
      });
    }

    case userServiceBaseUrl.host: {
      return authorize(redirectUrl.href)
        .then(result => {
          if (!result.success) {
            return result;
          }

          return handleRedirect(result.response.redirectUrl);
        });
    }

    default: {
      return buildSuccessResponse({
        redirectUrl: redirectUrl.href,
        external: true
      });
    }
  }
}

const signIn = (fields) => {
  const errors = {};
  // TODO the whole validation could have been implemented better, it does not check if username is even there
  const validationResult = validateSignInFields(fields, errors);
  if (!validationResult) {
    return Promise.resolve(
      buildErrorResponse({ fields: errors })
    );
  }

  const { username, password, csrfToken } = fields;
  const requestData = new URLSearchParams();
  requestData.append("username", username);
  requestData.append("password", password);

  return axiosInstance
    .post(
      "/sign-in",
      requestData,
      {
        withCredentials: true,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'X-XSRF-TOKEN': csrfToken
        }
      }
    ).then(response => {
      switch (response.status) {
        case 200: {
          const body = response.data;
          return handleRedirect(body.redirectUrl);
        }

        default: {
          return buildErrorResponse({ message: "Unable to process the request. Try again later."});
        }
      }
    }).catch(error => {
      if (error.response) {
        const response = error.response;
        switch (response.status) {
          case 401:
          case 403: {
            return buildErrorResponse({ message: "Invalid e-mail or password."});
          }
        }
      }

      return buildErrorResponse({ message: "Unable to process the request. Try again later." });
    });
};

const authorize = (request) => {
  if (!request) {
    return buildErrorResponse({ message: "Authorization request is empty."})
  }

  return axiosInstance
    .get(
      request,
      {
        withCredentials: true
      }
    ).then(response => {
      switch (response.status) {
        case 200: {
          const body = response.data;
          return buildSuccessResponse({
            redirectUrl: body.redirectUrl
          });
        }

        default: {
          return buildErrorResponse({ message: "Unable to process the request. Try again later."});
        }
      }
    }).catch(() => {
      return buildErrorResponse({message: "Unable to process the request. Try again later."});
    });
};

const validateConsentFields = (fields) => {
  const fieldsToValidate = { ...fields };
  delete fieldsToValidate.scopes;

  for (const key in fieldsToValidate) {
    if (!fields[key]) {
      return false;
    }
  }

  return true;
};

const consent = (fields) => {
  if (!validateConsentFields(fields)) {
    return Promise.resolve(
      buildErrorResponse({ message: "Unable to process the request. Try again later." })
    );
  }

  const { clientId, state, scopes } = fields;
  const requestData = new URLSearchParams();
  requestData.append("client_id", clientId);
  requestData.append("state", state);
  for (const scope of scopes) {
    requestData.append("scope", scope);
  }

  return axiosInstance
    .post(
      // TODO get url from config
      "/oauth2/authorize",
      requestData,
      {
        withCredentials: true,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        }
      }
    ).then(response => {
      switch (response.status) {
        case 200: {
          const body = response.data;
          return buildSuccessResponse({
            redirectUrl: body.redirectUrl ?? ""
          });
        }

        default: {
          return buildErrorResponse({ message: "Unable to process the request. Try again later." });
        }
      }
    }).catch(() => {
      return buildErrorResponse({ message: "Unable to process the request. Try again later." });
    });
};

const userService = {
  signIn,
  authorize,
  consent,
};

export { userService };