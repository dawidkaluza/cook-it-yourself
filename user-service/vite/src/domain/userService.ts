import {settings} from "../settings/settings.ts";
import {fetchApi} from "../api/fetch.ts";
import {ApiError} from "../api/ApiError.ts";
import {RedirectResponse, SignInRequest, SignInResponse, SignUpRequest, SignUpResponse} from "./dtos/user.ts";
import {InvalidCredentialsError, InvalidFieldsError} from "./errors/user.tsx";

const validate = (request: Record<string, any>) => {
  const errorFields: Record<string, string> = {};
  for (const key in request) {
    if (!request[key]) {
      errorFields[key] = "Field must not be empty.";
    }
  }
  return errorFields;
}

const isEmpty = (record: Record<string, string>) => {
  return !Object.keys(record).length;
}

const authorize = async (redirectUrl: string): Promise<RedirectResponse> => {
  return await fetchApi({
    endpoint: redirectUrl,
  });
};

const handleRedirect = async (redirectUrl?: string): Promise<SignInResponse> => {
  if (!redirectUrl) {
    return {
      redirectUrl: "",
      external: false,
    }
  }

  const appUrl = import.meta.env.BASE_URL;
  if (redirectUrl.startsWith(appUrl)) {
    return {
      redirectUrl,
      external: false,
    }
  }

  let userServiceUrl = settings.userServiceUrl || new URL(appUrl).origin;
  if (redirectUrl.startsWith(userServiceUrl)) {
    const redirectEndpoint = redirectUrl.slice(userServiceUrl.length);
    const authResponse = await authorize(redirectEndpoint);
    return handleRedirect(authResponse.redirectUrl);
  }

  return {
    redirectUrl: redirectUrl,
    external: true,
  }
};

const signIn = async (request: SignInRequest) => {
  const errorFields = validate(request);
  if (!isEmpty(errorFields)) {
    throw new InvalidCredentialsError("Invalid email or password.");
  }

  const requestData = new URLSearchParams();
  requestData.append("username", request.email);
  requestData.append("password", request.password);

  try {
    const body = await fetchApi<RedirectResponse>({
      endpoint: "/sign-in",
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: requestData
    });
    return handleRedirect(body.redirectUrl);
  } catch (error) {
    if (error instanceof ApiError) {
      switch (error.status) {
        case 401:
        case 403: {
          throw new InvalidCredentialsError("Invalid email or password.");
        }
      }
    }

    throw error;
  }
};

const signUp = async (request: SignUpRequest) => {
  const errorFields = validate(request);
  if (!isEmpty(errorFields)) {
    throw new InvalidFieldsError(errorFields);
  }

  try {
    return await fetchApi<SignUpResponse>({
      endpoint: "/user/sign-up",
      method: "POST",
      body: JSON.stringify(request)
    });
  } catch(error) {
    if (error instanceof ApiError) {
      switch(error.status) {
        case 409: {
          throw new InvalidFieldsError({
            email: "Email already exists"
          });
        }

        case 422: {
          const errorFields: Record<string, string> = {};
          const response: { fields: Array<{ name: string, message: string}> } = error.body;
          for (const field of response.fields) {
            errorFields[field.name] = field.message;
          }

          if (!isEmpty(errorFields)) {
            throw new InvalidFieldsError(errorFields);
          }
        }
      }
    }

    throw error;
  }
};

export const userService = { signIn, signUp };