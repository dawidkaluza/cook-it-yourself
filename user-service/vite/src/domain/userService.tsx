import {settings} from "../settings/settings.ts";
import {fetchApi} from "../api/fetch.ts";
import {ApiError} from "../api/ApiError.ts";
import {RedirectResponse, SignInRequest, SignInResponse} from "./dtos/user.ts";
import {InvalidCredentialsError} from "./errors/user.tsx";

const authorize = async (redirectUrl: string): Promise<RedirectResponse> => {
  return await fetchApi({
    endpoint: redirectUrl,
  });
};


const handleRedirect = async (redirectUrl?: string): Promise<SignInResponse> => {
  if (!redirectUrl) {
    return {
      redirectUrl: "",
      external: false,
    }
  }

  const appUrl = import.meta.env.BASE_URL;
  if (redirectUrl.startsWith(appUrl)) {
    return {
      redirectUrl,
      external: false,
    }
  }

  let userServiceUrl = settings.userServiceUrl;
  if (!userServiceUrl) {
    const publicPath = settings.publicPath;
    userServiceUrl = publicPath ? new URL(appUrl).href : appUrl;
  }

  if (redirectUrl.startsWith(userServiceUrl)) {
    const redirectEndpoint = redirectUrl.slice(userServiceUrl.length);
    const authResponse = await authorize(redirectEndpoint);
    return handleRedirect(authResponse.redirectUrl);
  }

  return {
    redirectUrl: redirectUrl,
    external: true,
  }
};

const signIn = async (request: SignInRequest) => {
  if (!request.email || !request.password) {
    throw new InvalidCredentialsError("Invalid username or password.");
  }

  const requestData = new URLSearchParams();
  requestData.append("username", request.email);
  requestData.append("password", request.password);

  try {
    const body = await fetchApi<RedirectResponse>({
      endpoint: "/sign-in",
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: requestData
    });
    return handleRedirect(body.redirectUrl);
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 401:
        case 403: {
          throw new InvalidCredentialsError("Invalid username or password.");
        }
      }
    }

    throw error;
  }
};

export const userService = { signIn };