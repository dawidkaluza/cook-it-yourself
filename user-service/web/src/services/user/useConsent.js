import {useCallback, useState} from "react";
import {userService} from "./userService";

const useConsent = () => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState("");
  const [error, setError] = useState("");

  const consent = useCallback(
    ({ clientId, state, scopes }) => {
      setLoading(false);
      setSuccess("");
      setError("");

      userService
        .consent({ clientId, state, scopes })
        .then(result => {
          if (!result.success) {
            const response = result.response;
            setError(response.message ?? "");
            return;
          }

          setSuccess("Consent approved successfully.");

          const redirectUrl = result.response.redirectUrl;
          if (redirectUrl) {
            setTimeout(() => {
              window.location.assign(redirectUrl);
            }, 800);
          }
        })
        .finally(() => {
          setLoading(false);
        });

    }, []
  );

  return { consent, loading, success, error };
};

export { useConsent };