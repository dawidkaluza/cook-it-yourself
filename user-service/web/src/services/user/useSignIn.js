import {useCallback, useState} from "react";
import {userService} from "./userService";
import {useCookies} from "react-cookie";
import {useNavigate} from "react-router-dom";

const useSignIn = () => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState("");
  const [error, setError] = useState("");
  const [fieldsErrors, setFieldsErrors] = useState({});
  const [cookies] = useCookies();
  const navigate = useNavigate();

  const signIn = useCallback(
    ({username, password}) => {
      setLoading(true);
      setSuccess("");
      setError("");
      setFieldsErrors({});

      // TODO what if cookie does not exist?
      const csrfToken = cookies['XSRF-TOKEN'];

      return userService
        .signIn({ username, password, csrfToken })
        .then(result => {
          if (!result.success) {
            const response = result.response;
            setError(response.message ?? "");
            setFieldsErrors(response.fields ?? {});
            return result;
          }

          setSuccess("Signing in proceeded successfully.");

          const response = result.response;
          if (response.external) {
            setTimeout(() => {
              window.location.assign(response.redirectUrl);
            }, 800);
          } else {
            setTimeout(() => {
              navigate(response.redirectUrl);
            }, 800);
          }
        })
        .finally(() => {
          setLoading(false);
        });
    }, []
  );

  return { signIn, loading, success, error, fieldsErrors };
};

export { useSignIn };