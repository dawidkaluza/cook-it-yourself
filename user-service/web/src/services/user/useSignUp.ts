import {useCallback, useState} from "react";
import {useCookies} from "react-cookie";
import {userService} from "./userService";
import {useNavigate} from "react-router-dom";

const useSignUp = () => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState("");
  const [error, setError] = useState("");
  const [fieldsErrors, setFieldsErrors] = useState({});
  const [cookies] = useCookies();
  const navigate = useNavigate();

  const signUp = useCallback(
    ({ email, password, name }) => {
      setLoading(true);
      setSuccess("");
      setError("");
      setFieldsErrors({});

      userService
        .signUp({ email, password, name })
        .then(result => {
          if (!result.success) {
            const response = result.response;
            setError(response.message ?? "");
            setFieldsErrors(response.fields ?? {});
            return;
          }

          setSuccess("Signing up proceeded successfully. You will be redirected to sign in page...");
          setTimeout(() => {
            navigate("/sign-in");
          }, 800);
        })
        .finally(() => {
          setLoading(false);
        })
    }, []
  );

  return { signUp, loading, success, error, fieldsErrors };
};

export { useSignUp }