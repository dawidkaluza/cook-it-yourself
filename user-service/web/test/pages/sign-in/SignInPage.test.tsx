import {afterAll, beforeAll, beforeEach, describe, expect, Mock} from "vitest";
import {SignInPage} from "../../../src/pages/sign-in/SignInPage.tsx";
import {render, screen} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import {userEvent} from "@testing-library/user-event";
import {userService} from "../../../src/domain/userService.ts";
import {InvalidCredentialsError} from "../../../src/domain/errors/user.tsx";

vi.mock("react", async function <T>(importOriginal: () => Promise<T>) {
  const original = await importOriginal();
  return {
    ...original,
    useId: () => "1234",
  };
});

const navigateMock = vi.fn();
vi.mock("react-router-dom", async function <T>(importOriginal: () => Promise<T>) {
  const original = await importOriginal();
  return {
    ...original,
    useNavigate: () => navigateMock,
  };
});

vi.mock("../../../src/domain/userService.ts", () => {
  return {
    userService: {
      signIn: vi.fn(),
    }
  };
});

beforeAll(() => {
  vi.stubGlobal("console", {
    error: vi.fn()
  });

  vi.stubGlobal("location", {
    assign: vi.fn()
  });
});

beforeEach(() => {
  vi.clearAllMocks();
});

afterAll(() => {
  vi.unstubAllGlobals();
});

describe("SignInPage component", () => {
  test("render", () => {
    // Given, when
    const component = render(
      <MemoryRouter>
        <SignInPage />
      </MemoryRouter>
    );

    // Then
    const emailInputEl = screen.getByLabelText("E-mail");
    expect(emailInputEl).toBeDefined();
    const passwordInputEl = screen.getByLabelText("Password");
    expect(passwordInputEl).toBeDefined();
    const submitBtnEl = screen.getByRole("button", { name: "Sign in" });
    expect(submitBtnEl).toBeDefined();
    const signUpLinkEl = screen.getByText(/sign up/i);
    expect(signUpLinkEl).toBeDefined();
    expect(signUpLinkEl.getAttribute("href")).toBe("/sign-up");

    expect(component.container).toMatchSnapshot();
  });

  test("type email and password, inputs change", async () => {
    // Given
    const user = userEvent.setup();

    const component = render(
      <MemoryRouter>
        <SignInPage />
      </MemoryRouter>
    );

    // When
    const emailInputEl = screen.getByLabelText("E-mail");
    await user.click(emailInputEl);
    await user.keyboard("dawid@gmail.com");

    const passwordInputEl = screen.getByLabelText("Password");
    await user.click(passwordInputEl);
    await user.keyboard("passwd");

    // Then
    expect(emailInputEl.getAttribute("value")).toBe("dawid@gmail.com");
    expect(passwordInputEl.getAttribute("value")).toBe("passwd");

    expect(component.container).toMatchSnapshot();
  });

  test.each([
    {
      error: new InvalidCredentialsError("Invalid email or password."),
      expectedMessage: /invalid email or password/i,
      expectedLog: false,
    },
    {
      error: new Error("Network error."),
      expectedMessage: /unable to process the request/i,
      expectedLog: true,
    },
  ])("submit, $error.message error thrown, message matching $expectedMessage shown", async ({ error, expectedMessage, expectedLog }) => {
    // Given
    (userService.signIn as Mock).mockRejectedValue(error);

    const user = userEvent.setup();
    const component = render(
      <MemoryRouter>
        <SignInPage />
      </MemoryRouter>
    );

    const emailInputEl = screen.getByLabelText("E-mail");
    await user.click(emailInputEl);
    await user.keyboard("dawid@gmail.com");

    const passwordInputEl = screen.getByLabelText("Password");
    await user.click(passwordInputEl);
    await user.keyboard("invalidpasswd");

    // When
    const submitBtnEl = screen.getByRole("button", { name: "Sign in" });
    await user.click(submitBtnEl);

    // Then
    const errorMsgEl = screen.getByText(expectedMessage);
    expect(errorMsgEl).toBeDefined();

    if (expectedLog) {
      expect(console.error).toHaveBeenCalledWith(
        expect.anything(),
        error
      );
    }

    expect(component.container).toMatchSnapshot();
  });

  test.each([
    {
      signInResponse: {
        redirectUrl: "http://user-service/web/consent",
        external: false,
      }
    },
    {
      signInResponse: {
        redirectUrl: "http://api-gateway/oauth2?code=123xyz",
        external: true,
      }
    },
  ])("submit with valid username and password, redirect as expected", async ({ signInResponse }) => {
    // Given
    (userService.signIn as Mock).mockResolvedValue(signInResponse);

    const user = userEvent.setup();
    const component = render(
      <MemoryRouter>
        <SignInPage />
      </MemoryRouter>
    );

    const emailInputEl = screen.getByLabelText("E-mail");
    await user.click(emailInputEl);
    await user.keyboard("dawid@gmail.com");

    const passwordInputEl = screen.getByLabelText("Password");
    await user.click(passwordInputEl);
    await user.keyboard("passwd");

    // When
    const submitBtnEl = screen.getByRole("button", { name: "Sign in" });
    await user.click(submitBtnEl);

    // Then
    const successMsgEl = screen.getByText(/signing in proceeded successfully/i);
    expect(successMsgEl).toBeDefined();

    if (signInResponse.external) {
      await expect.poll(() => window.location.assign).toHaveBeenCalledWith(signInResponse.redirectUrl);
    } else{
      await expect.poll(() => navigateMock).toHaveBeenCalledWith(signInResponse.redirectUrl);
    }

    expect(component.container).toMatchSnapshot();
  });
});