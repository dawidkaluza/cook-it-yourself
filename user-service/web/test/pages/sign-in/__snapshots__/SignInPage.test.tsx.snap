// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignInPage component > render 1`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value=""
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value=""
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
  </form>
</div>
`;

exports[`SignInPage component > submit with valid username and password, redirect as expected 1`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value="dawid@gmail.com"
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value="passwd"
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
    <p
      class="text-green-600"
    >
      Signing in proceeded successfully.
    </p>
  </form>
</div>
`;

exports[`SignInPage component > submit with valid username and password, redirect as expected 2`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value="dawid@gmail.com"
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value="passwd"
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
    <p
      class="text-green-600"
    >
      Signing in proceeded successfully.
    </p>
  </form>
</div>
`;

exports[`SignInPage component > submit, 'Invalid email or password.' error thrown, message matching /invalid email or password/i shown 1`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value="dawid@gmail.com"
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value="invalidpasswd"
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
    <p
      class="text-red-600"
    >
      Invalid email or password.
    </p>
  </form>
</div>
`;

exports[`SignInPage component > submit, 'Network error.' error thrown, message matching /unable to process the request/i shown 1`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value="dawid@gmail.com"
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value="invalidpasswd"
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
    <p
      class="text-red-600"
    >
      Unable to process the request. Try again later.
    </p>
  </form>
</div>
`;

exports[`SignInPage component > type email and password, inputs change 1`] = `
<div>
  <form
    class="flex flex-col items-center max-w-sm mx-auto mt-8"
  >
    <h1
      class="mb-4 text-center text-2xl"
    >
      Sign in
    </h1>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="email_1234"
      >
        E-mail
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="email_1234"
        name="email"
        value="dawid@gmail.com"
      />
    </div>
    <div
      class="w-full mb-3 px-8"
    >
      <label
        class="block mb-1 font-medium"
        for="password_1234"
      >
        Password
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="password_1234"
        name="password"
        type="password"
        value="passwd"
      />
    </div>
    <button
      class="bg-blue-500 hover:bg-blue-600 text-white font-bold py-2 px-6 rounded disabled:opacity-75 disabled:hover:bg-blue-500 mb-5"
      type="submit"
    >
      Sign in
    </button>
    <p>
      Don't have an account? 
       
      <a
        class="font-bold text-blue-500 hover:text-blue-700"
        href="/sign-up"
      >
        Sign up.
      </a>
    </p>
  </form>
</div>
`;
